# Exploring Ebay Car Sales Data

This project will involve the exploration and analysis of data collected from eBay Kleinanzeigen - a classifieds portion of the German eBay website. 

The [dataset](https://www.kaggle.com/orgesleka/used-cars-database/data) used is sourced from kaggle but was originally scraped form the website by dataquest and then modified slightly to appear like a more realistic dataset that requires more cleaning and pre-processing. 

First, the datset will be imported into a dataframe and explored.

In [1]:
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [2]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

As can be seen above, the dataset contains a mixture of object types and int64 types. Some of the columns are represented as objects when they should be integers. This will be handled in the following section of the code.

## Cleaning Column Labels

The columns are in CamelCase format. However, it would be more applicable for them to be in snake_case format. Therefore, the code below will make this conversion and make direct modifications to column names where needed.

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
autos.rename({"yearOfRegistration":"registration_year"}, axis=1, inplace=True)
autos.rename({"monthOfRegistration":"registration_month"}, axis=1, inplace=True)
autos.rename({"notRepairedDamage":"unrepaired_damage"}, axis=1, inplace=True)
autos.rename({"dateCreated":"ad_created"}, axis=1, inplace=True)
autos.rename({"powerPS":"power_ps"}, axis=1, inplace=True)
autos.rename({"abtest":"ab_test"}, axis=1, inplace=True)

for column in autos.columns:
    
    old_column = column
    
    for char in column:
        if char == char.upper() and char != '_':
            column = column.replace(char, ('_' + char.lower()))
            
    autos.rename({old_column:column}, axis=1, inplace=True)

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Cleaning Data

As stated above, some columns are of the wrong type. Also, much of the data contains the units in the rows instead of the header. This makes analysing such columns difficult as they cannot be expressed numerically. The `odometer` and `price` columns will be modified so that the header contains the units and the data is of int64 type.

In [6]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 50000 non-null object
ab_test               50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gearbox               47320 non-null object
power_ps              50000 non-null int64
model                 47242 non-null object
odometer              50000 non-null object
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
nr_of_pictures        50000 non-null int64
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: int64(5)

In [7]:
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","")
autos["odometer"] = autos["odometer"].astype(int)
autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)

autos["price"] = autos["price"].str.replace("$","").str.replace(",","")
autos["price"] = autos["price"].astype(int)
autos.rename({"price":"price_$"}, axis=1, inplace=True)

In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price_$  ab_test vehicle_type  registration_year  \
0  privat    Angebot     5000  control          bus               2004   
1  privat    Angebot     8500  control    limousine               1997   
2  privat    Angebot     8990     test    limousine               2009   
3  privat    Angebot     4350  control   kleinwagen               2007   
4  privat    Angebot     1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Now that the columns have been cleaned, it will be useful to analyse them for any outliers or unexpected trends in the data.

In [9]:
print(autos["odometer_km"].unique().shape)
print(autos["odometer_km"].describe())

(13,)
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


In [10]:
autos["odometer_km"].value_counts().sort_index()

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

The analysis of the `odometer_km` column above shows no obvious outliers or peculiar trends. It can be said, however, that there is a trend towards more high mileage cars being listed on the site.

Next, the `price` column will be analysed in the same manner.

In [11]:
print(autos["price_$"].unique().shape)
print(autos["price_$"].describe())

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_$, dtype: float64


In [12]:
autos["price_$"].value_counts().sort_index().head(30)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
40       6
45       4
47       1
49       4
50      49
55       2
59       1
60       9
65       5
66       1
Name: price_$, dtype: int64

In [13]:
autos["price_$"].value_counts().sort_index(ascending=False).head(30)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
Name: price_$, dtype: int64

Inspecting the above two code cells, it can be seen that on the lower end of the price spectrum, there are a disproportionate number of listings with a price of \$0. There then seems to be a larger amount of listings at \$50 and up. This price seems to realisticlly be the cheapest that a vehicle might sell for. Likewise, on the upper end of the price spectrum, there is a huge jump between \$350,000, \$999,990 and upwards. This small sample of listings with a huge list price will skew the dataset. Therefore, all listings below \$50 will be removed from the dataset and likewise, all listings above \$350,000 will be removed for the dataset.

In [14]:
autos = autos[autos["price_$"].between(50,350000)]
autos["price_$"].describe()

count     48347.000000
mean       5915.456099
std        9071.625794
min          50.000000
25%        1250.000000
50%        3000.000000
75%        7499.000000
max      350000.000000
Name: price_$, dtype: float64

## Exploring the Date Columns

Looking at the dataset, there are 5 date-related columns:

- `date_crawled`: The date that the listing was added by the crawler.
- `registration_year`: The year the vehicle was registered.
- `registration_month`: The month the vehicle was registered.
- `ad_created`: The datetime that the listing was made.
- `last_seen`: The datetime that the crawler last saw the listing.

`registration_year` and `registration_month` are both numeric so can therefore be analysed simply using the `.describe()` method. Conversely, the remaining three columns are represented as strings so will require some more data processing to analyse as seen below.

First, the format of the 3 columns represented as strings will be explored:

In [15]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [16]:
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025358
2016-03-06    0.014065
2016-03-07    0.036031
2016-03-08    0.033260
2016-03-09    0.033053
2016-03-10    0.032225
2016-03-11    0.032556
2016-03-12    0.036983
2016-03-13    0.015658
2016-03-14    0.036672
2016-03-15    0.034294
2016-03-16    0.029495
2016-03-17    0.031584
2016-03-18    0.012907
2016-03-19    0.034749
2016-03-20    0.037789
2016-03-21    0.037293
2016-03-22    0.032949
2016-03-23    0.032267
2016-03-24    0.029433
2016-03-25    0.031501
2016-03-26    0.032267
2016-03-27    0.031088
2016-03-28    0.034894
2016-03-29    0.034108
2016-03-30    0.033715
2016-03-31    0.031832
2016-04-01    0.033735
2016-04-02    0.035535
2016-04-03    0.038575
2016-04-04    0.036528
2016-04-05    0.013051
2016-04-06    0.003165
2016-04-07    0.001386
Name: date_crawled, dtype: float64

From above, it can be seen that there many new listings being crawled over time. Likewise, there seems to be no irregularities in the data of the column.

In [17]:
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033135
2016-03-10    0.031936
2016-03-11    0.032887
2016-03-12    0.036796
2016-03-13    0.017023
2016-03-14    0.035307
2016-03-15    0.034025
2016-03-16    0.029992
2016-03-17    0.031253
2016-03-18    0.013589
2016-03-19    0.033632
2016-03-20    0.037851
2016-03-21 

There is a large variety in the date in which the listings were created. It can be seen that there are far more newer ads than older ads. This makes sense since you would expect most of the older ads to have sold. The oldest ad on the site is around 10 months old.

In [18]:
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001076
2016-03-06    0.004323
2016-03-07    0.005419
2016-03-08    0.007343
2016-03-09    0.009639
2016-03-10    0.010611
2016-03-11    0.012390
2016-03-12    0.023745
2016-03-13    0.008873
2016-03-14    0.012638
2016-03-15    0.015885
2016-03-16    0.016464
2016-03-17    0.028089
2016-03-18    0.007322
2016-03-19    0.015802
2016-03-20    0.020622
2016-03-21    0.020601
2016-03-22    0.021408
2016-03-23    0.018595
2016-03-24    0.019774
2016-03-25    0.019174
2016-03-26    0.016775
2016-03-27    0.015575
2016-03-28    0.020849
2016-03-29    0.022318
2016-03-30    0.024738
2016-03-31    0.023828
2016-04-01    0.022876
2016-04-02    0.024883
2016-04-03    0.025172
2016-04-04    0.024510
2016-04-05    0.124972
2016-04-06    0.221627
2016-04-07    0.132087
Name: last_seen, dtype: float64

The `last_seen` data is skewed towards the most recent 3 days in the dataset. This sudden increase is likely due to the crawler not updating and the items seen in the last 3 days are active and yet to be sold.

Next, the numercially represented `registration_year` column will be explored.

In [19]:
autos["registration_year"].describe()

count    48347.000000
mean      2004.718845
std         87.786604
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

As can be seen above, the minimum value of 1000 is obviously incorrect since cars have not been in production until at earliest, the early 1900s. Likewise, there should not be a listing with a registration year greater than 2016 since that is neither possible. 

Therefore, all listings with a registration year before 1900 and greater than 2016 will be removed from the dataset.

In [20]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].describe()

count    46473.000000
mean      2002.931315
std          7.128802
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

After removing the incorrect data, it can be seen that the minimum registration year is more relaistically stated as 1910 and the maximum year is now 2016. It is also important to state that this has only removed around 200 listings from the dataset of over 45000.

## Exploring Price by Brand

In [21]:
brands_dist = autos["brand"].value_counts(normalize=True).sort_values(ascending=False)
brands_dist

volkswagen        0.211327
bmw               0.109956
opel              0.107310
mercedes_benz     0.096508
audi              0.086609
ford              0.069976
renault           0.047189
peugeot           0.029910
fiat              0.025714
seat              0.018247
skoda             0.016418
nissan            0.015321
mazda             0.015256
smart             0.014180
citroen           0.014051
toyota            0.012760
hyundai           0.010049
sonstige_autos    0.009618
volvo             0.009102
mini              0.008779
mitsubishi        0.008177
honda             0.007854
kia               0.007101
alfa_romeo        0.006649
porsche           0.006047
suzuki            0.005939
chevrolet         0.005702
chrysler          0.003529
dacia             0.002647
daihatsu          0.002518
jeep              0.002281
subaru            0.002109
land_rover        0.002109
saab              0.001657
jaguar            0.001571
daewoo            0.001485
trabant           0.001377
r

It would not be useful to analyse all of the brands in the data as many of the brands make up such a small percentage of the listings. Therfore, only brands with a share above 1% will be analysed going forward.

In [22]:
popular_brands = brands_dist[brands_dist > .01].index
popular_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai'],
      dtype='object')

In [23]:
popular_brands_prices = {}

for brand in popular_brands:
    selected_brand = autos[autos["brand"] == brand]
    popular_brands_prices[brand] = int(selected_brand["price_$"].mean())
    
popular_brands_prices

{'volkswagen': 5424,
 'bmw': 8376,
 'opel': 2996,
 'mercedes_benz': 8663,
 'audi': 9373,
 'ford': 3762,
 'renault': 2483,
 'peugeot': 3100,
 'fiat': 2818,
 'seat': 4423,
 'skoda': 6393,
 'nissan': 4750,
 'mazda': 4112,
 'smart': 3591,
 'citroen': 3784,
 'toyota': 5167,
 'hyundai': 5376}

Of all the most popular brands listed on the site:

- Audi, BMW and Mercedes-Benz are the most expensive.
- Volkswagen, Seat, Skoda, Hyundai, Nissan and Mazda are less expensive.
- Opel, Ford, Renault, Peugeot, Fiat, Smart and Citroen are the cheapest.

## Exploring Mileage by Brand

In [29]:
mean_price = pd.Series(popular_brands_prices).sort_values(ascending=False)
mean_price

audi             9373
mercedes_benz    8663
bmw              8376
skoda            6393
volkswagen       5424
hyundai          5376
toyota           5167
nissan           4750
seat             4423
mazda            4112
citroen          3784
ford             3762
smart            3591
peugeot          3100
opel             2996
fiat             2818
renault          2483
dtype: int64

In [30]:
popular_brands_mileage = {}

for brand in popular_brands:
    selected_brand = autos[autos["brand"] == brand]
    popular_brands_mileage[brand] = int(selected_brand["odometer_km"].mean())
    
mean_mileage = pd.Series(popular_brands_mileage).sort_values(ascending=False)


mean_mileage_df = pd.DataFrame(mean_mileage, columns=['mean_mileage'])
mean_mileage

bmw              132651
mercedes_benz    130923
opel             129440
audi             129254
volkswagen       128771
renault          128144
peugeot          127122
mazda            124464
ford             124295
seat             121303
citroen          119647
nissan           118370
fiat             117066
toyota           115944
skoda            110884
hyundai          106541
smart             99590
dtype: int64

In [32]:
popular_brands_df = pd.DataFrame(mean_price, columns=['mean_price'])
popular_brands_df['mean_mileage'] = mean_mileage

popular_brands_df                                  

mean_price  mean_mileage
audi                 9373        129254
mercedes_benz        8663        130923
bmw                  8376        132651
skoda                6393        110884
volkswagen           5424        128771
hyundai              5376        106541
toyota               5167        115944
nissan               4750        118370
seat                 4423        121303
mazda                4112        124464
citroen              3784        119647
ford                 3762        124295
smart                3591         99590
peugeot              3100        127122
opel                 2996        129440
fiat                 2818        117066
renault              2483        128144

As can be seen in the above dataframe, there is little variability between the top brands based on mileage. Although, there is a slight trend for the more expensive brands to have a slightly higher mileage than the cheaper brands.